src/java.base/share/classes/java/util/UUID.java

```diff
@@ -99,24 +104,190 @@ public final class UUID implements java.io.Serializable, Comparable<UUID> {
     /*
      * The random number generator used by this class to create random
      * based UUIDs. In a holder class to defer initialization until needed.
+     * <p>
+     * The implementation buffers the reads to avoid bashing SecureRandom with
+     * small requests. It also maintains a SecureRandom per buffer to alleviate
+     * scalability bottlenecks when reading from a single (synchronized) SecureRandom.
      */
-    private static class Holder {
-        static final SecureRandom numberGenerator = new SecureRandom();
+    private static final class RandomUUID {
+        static final String PROP_NAME_PRNG_NAME = "java.util.UUID.prngName";
+        static final String PROP_NAME_BUF_COUNT = "java.util.UUID.buffersCount";
+
+        static final String PRNG_NAME;
+        static final int BUFS_COUNT;
+        static final Buffer[] BUFS;
+
+        private static final int roundPowerOfTwo(int x) {
+            int n = -1 >>> Integer.numberOfLeadingZeros(x - 1);
+            return (n < 0) ? 1 : (n + 1);
+        }
+
+        static {
+            try {
+                PRNG_NAME = System.getProperty(PROP_NAME_PRNG_NAME, null);
+                try {
+                    newRandom();
+                } catch (Exception e) {
+                    throw new IllegalArgumentException(PROP_NAME_PRNG_NAME + " is incorrect", e);
+                }
+                int bufCount = Integer.getInteger(PROP_NAME_BUF_COUNT, Runtime.getRuntime().availableProcessors());
+                if (bufCount < 1) {
+                    throw new IllegalArgumentException(PROP_NAME_BUF_COUNT + " is out of range");
+                }
+                BUFS_COUNT = roundPowerOfTwo(bufCount);
+                BUFS = new Buffer[BUFS_COUNT];
+            } catch (Exception e) {
+                throw new InternalError(e);
+            }
+        }
+
+        static SecureRandom newRandom() {
+            if (PRNG_NAME == null) {
+                return new SecureRandom();
+            } else {
+                try {
+                    return SecureRandom.getInstance(PRNG_NAME);
```

Comment:
Part of the change here is that the RNG algorithm is configurable via a system property. The naming of the naming (java.util.UUID.prngName) hints that you might want it to be a standard system property but maybe it's for testing?

Reply:
Yes, I used them for testing, but left them in, because it would be convenient to have them around for field tuning and diagnostics. This would require a CSR, right?

Reply:
I don't think there is a strong case for adding the system properties. They just add to the surface area that has to be maintained and add very little value except in very rare cases.  If someone needs very special treatment they can build it outside of UUID and use the new UUID(long, long) constructor.

Reply:
That sounds okay to me. It simplifies the whole thing and does not require me to implement more test and have a CSR discussion about the options too :) Removed in the new commit.